pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -95,49 +95,50 @@ public class ModelInterpreter implements Serializable {
 
-                            script.stage(thisStage.name) {
-                                withEnvBlock(thisStage.getEnvVars()) {
-                                    if (firstError == null) {
-                                        inDeclarativeAgent(thisStage.agent) {
-                                            toolsBlock(thisStage.agent ?: root.agent, thisStage.tools) {
-                                                try {
-                                                    catchRequiredContextForNode(root.agent) {
-                                                        setUpDelegate(thisStage.steps.closure).call()
-                                                    }.call()
-                                                } catch (Exception e) {
-                                                    script.echo "Error in stages execution: ${e.getMessage()}"
-                                                    script.getProperty("currentBuild").result = Result.FAILURE
-                                                    if (firstError == null) {
-                                                        firstError = e
-                                                    }
-                                                } finally {
-                                                    // And finally, run the post stage steps.
-                                                    List<Closure> postClosures = thisStage.satisfiedPostStageConditions(root, script.getProperty("currentBuild"))
-                                                        catchRequiredContextForNode(thisStage.agent != null ? thisStage.agent : root.agent, false) {
-                                                            if (postClosures.size() > 0) {
-                                                                script.echo("Post stage")
-                                                                //TODO should this be a nested stage instead?
-                                                                try {
-                                                                    for (int ni = 0; ni < postClosures.size(); ni++) {
-                                                                        setUpDelegate(postClosures.get(ni)).call()
-                                                                    }
-                                                                } catch (Exception e) {
-                                                                    script.echo "Error in stage post: ${e.getMessage()}"
-                                                                    script.getProperty("currentBuild").result = Result.FAILURE
-                                                                    if (firstError == null) {
-                                                                        firstError = e
+                                runStageOrNot(thisStage, firstError) {
```

Comment:
:bug: Flip with the line below. We always want to have the `stage` even if we don't do anything in it. This is so that we have a consistent, predictable execution model.